### PR TITLE
Add an option to DevTools to enable double-logging

### PR DESCRIPTION
**Pull request by Jchinonso**
## Summary

 We currently plan to suppress double-logging in development by overriding console in #18547. However, this can be confusing for some scenarios and double logging may be preferable in them. This change adds a toggle to DevTools that would allow to re-enable it on this screen:

90825360-b7372000-e330-11ea-9128-803d35d27e07

## Test Plan

Screenshot from 2020-09-04 00-18-20

### CODE REVIEW

**Comment by bvaughn** [comment]
Hi @Jchinonso. What's the purpose of this change?

**Comment by Jchinonso** [comment]
@bvaughn  We currently plan to suppress double-logging in development by overriding console in https://github.com/facebook/react/pull/18547.
However,  this can be confusing for some scenarios and double logging may be preferable in them.
This changes adds a toggle to DevTools that would allow to re-enable it on this screen:

**Comment by Jchinonso** [comment]
> Hi @Jchinonso. What's the purpose of this change?

@bvaughn  https://github.com/facebook/react/pull/19710#issue-474864087

**Comment by bvaughn** [comment]
There are two things that override the console:
1. React itself, before double rendering (in strict mode, DEV only) (see #18547)
2. DevTools, before re-rendering a component to inspect its hooks (see #15726)

Such a setting as the one you're adding here could be used to control the 2nd (DevTools) but I don't think anyone is asking for us to stop disabling the console in that case– so I assume you're targeting the first (React's disabling of the console, added in #18547).

That code wasn't written to be turned on or off externally (by something like DevTools). It's built into React. Allowing DevTools to control it (via a new setting like the one you're proposing) would require updating React as well.

Not sure how I feel about this feature :smile: but I'll be happy to look over the code when it's ready.

**Comment by Jchinonso** [comment]
@bvaughn thanks for the clarification, I saw the patching on the console for both devtools and react and was confused on how to approached this. Now i know i will take the React route

File: packages/shared/ConsolePatchingDev.js [hunk 1]
```diff
@@ -20,71 +20,76 @@ let prevError;
 let prevGroup;
 let prevGroupCollapsed;
 let prevGroupEnd;
+const enableDoubleLogging = global.__REACT_DEVTOOLS_ENABLE_DOUBLE_LOGGING__;
```
Comment (gaearon) [inline packages/shared/ConsolePatchingDev.js:23]: This is only read once, right? So what if the user switches midway through the session? Seems like it would be nice to for this setting to take effect every time we get to `disabledDepth === 0`. (But we should not change it midway through an existing log.)

Comment (gaearon) [inline packages/shared/ConsolePatchingDev.js:23]: What is `global` here? I don't think this would work in the browser. I think you'll want to do something like

```js
import {canUseDOM} from 'shared/ExecutionEnvironment';
```

and then if `canUseDOM` is true, read if from `window`.

File: packages/react-devtools-shared/src/devtools/views/Settings/DebuggingSettings.js [hunk 2]
```diff
@@ -35,7 +37,18 @@ export default function DebuggingSettings(_: {||}) {
...
+      <div className={styles.Setting}>
+        <label>
+          <input
+            type="checkbox"
+            checked={enableDoubleLogging}
+            onChange={({currentTarget}) =>
+              setEnableDoubleLogging(currentTarget.checked)
+            }
+          />{' '}
+          Enable double logging
```
Comment (gaearon) [inline packages/react-devtools-shared/src/devtools/views/Settings/DebuggingSettings.js:49]: This is not a very clear name for an option. A React user wouldn't know what this means.

I think instead it should be something like "Suppress console during development-only second render pass" and true by default. I.e. invert the option.

**Review by gaearon — CHANGES_REQUESTED** [review]
Thanks for working on this! I think we need to change a few things.

File: packages/shared/ConsolePatchingDev.js [hunk 3]
```diff
@@ -20,13 +21,16 @@ let prevError;
 let prevGroup;
 let prevGroupCollapsed;
 let prevGroupEnd;
+const {
+  __REACT_DEVTOOLS_ENABLE_DOUBLE_LOGGING__: enableDoubleLogging,
```
Comment (gaearon) [inline packages/shared/ConsolePatchingDev.js:25]: This still reads the value only once at startup. So switching it midway would not work. I think you want to update this value each time we enter `disableLogs` and `disabledDepth === 0`.

File: packages/react-devtools-shared/src/devtools/views/Settings/DebuggingSettings.js [hunk 4]
```diff
@@ -35,7 +37,18 @@ export default function DebuggingSettings(_: {||}) {
...
         </label>
       </div>
-
+      <div className={styles.Setting}>
+        <label>
+          <input
+            type="checkbox"
+            checked={enableDoubleLogging}
+            onChange={({currentTarget}) =>
+              setEnableDoubleLogging(currentTarget.checked)
```
Comment (gaearon) [inline packages/react-devtools-shared/src/devtools/views/Settings/DebuggingSettings.js:46]: I think you'll want `setEnableDoubleLogging(!currentTarget.checked)` and `checked={!enableDoubleLogging}` since the meaning is the opposite. And then keep `false` as the default.

Alternatively, you can change all occurrences of the key to change the meaning as well.

Comment (Jchinonso) [reply packages/react-devtools-shared/src/devtools/views/Settings/DebuggingSettings.js:46]: @gaearon i don't understand what you mean, don't you think it will change the course of things?

Comment (Jchinonso) [reply packages/react-devtools-shared/src/devtools/views/Settings/DebuggingSettings.js:46]: Alright, still looking at it by the old name `Enable double logging`

File: packages/shared/ConsolePatchingDev.js (hunk 3 above)
Comment (Jchinonso) [reply packages/shared/ConsolePatchingDev.js:25]: Please, can i get your input on this. I tried but it seems  `getEnableDoubleLogging()` doesn't update

Comment (Jchinonso) [reply packages/shared/ConsolePatchingDev.js:25]: @gaearon

Comment (Jchinonso) [reply packages/shared/ConsolePatchingDev.js:25]: i tried this approach, 


```export function disableLogs(): void {
  if (__DEV__) {
    if (disabledDepth === 0) {
      /* eslint-disable react-internal/no-production-logging */
      console.log(another, '====>');
      try {
        if (!another) {
          patchedConsole();
        }
      } finally {
        another = suppressDoubleLogging;
      }
    }
    /* eslint-enable react-internal/no-production-logging */
  }
  disabledDepth++;
}```

Comment (Jchinonso) [reply packages/shared/ConsolePatchingDev.js:25]: @gaearon 

```
function patchConsole() {
  prevLog = console.log;
  prevInfo = console.info;
  prevWarn = console.warn;
  prevError = console.error;
  prevGroup = console.group;
  prevGroupCollapsed = console.groupCollapsed;
  prevGroupEnd = console.groupEnd;
  // https://github.com/facebook/react/issues/19099
  const props = {
    configurable: true,
    enumerable: true,
    value: disabledLog,
    writable: true,
  };
  // $FlowFixMe Flow thinks console is immutable.
  Object.defineProperties(console, {
    info: props,
    log: props,
    warn: props,
    error: props,
    group: props,
    groupCollapsed: props,
    groupEnd: props,
  });
}

Comment (Jchinonso) [reply packages/shared/ConsolePatchingDev.js:25]:
```
export function getSuppressDoubleLogging(): boolean {
  try {
    const raw = localStorageGetItem(LOCAL_STORAGE_SHOULD_ENABLE_DOUBLE_LOGGING);
    if (raw != null) {
      return JSON.parse(raw);
    }
  } catch (error) {}
  return false;
}

File: fixtures/devtools/standalone/index.html [hunk 5]
```diff
@@ -251,63 +253,65 @@ <h1>List</h1>
...
+      );
+    }
+
+    ReactDOM.render(<App />, document.getElementById("root"));
+  </script>
+</body>
 
-      ReactDOM.render(<App />, document.getElementById("root"));
-    </script>
-  </body>
```
Comment (bvaughn) [inline fixtures/devtools/standalone/index.html:312]: Why all of the whitespace changes to this fixtures file? Seems unrelated to the PR.

Comment (Jchinonso) [reply fixtures/devtools/standalone/index.html:312]: Yeah, i skip that i shouldn't have committed.

**Comment by zhangenming** [comment]
too impatient to wait

**Comment by eps1lon** [comment]
Replaced by https://github.com/facebook/react/pull/22030
